onnxruntime/contrib_ops/cuda/grid_sample.cc

```diff
@@ -0,0 +1,96 @@
+namespace contrib {
+namespace cuda {
+
+#define REGISTER_KERNEL_TYPED(T)                                  \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
+      GridSample,                                                 \
+      kMSDomain,                                                  \
+      1,                                                          \
+      T,                                                          \
+      kCudaExecutionProvider,                                     \
+      (*KernelDefBuilder::Create())                               \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      GridSample<T>);
+
+REGISTER_KERNEL_TYPED(float)
+
+template <typename T>
+GridSample<T>::GridSample(const OpKernelInfo& info) : CudaKernel(info) {
+  std::string mode_str = info.GetAttrOrDefault<std::string>("mode", "bilinear");
+  std::string padding_mode_str = info.GetAttrOrDefault<std::string>("padding_mode", "zeros");
+  align_corners_ = static_cast<bool>(info.GetAttrOrDefault<int64_t>("align_corners", 0));
+  ORT_ENFORCE(mode_str == "bilinear" || mode_str == "nearest" || mode_str == "bicubic", 
+      "mode \"", mode_str, "\" not supported, expect bilinear, nearest or bicubic");
+  ORT_ENFORCE(padding_mode_str == "zeros" || padding_mode_str == "border" || padding_mode_str == "reflection",
+      "padding_mode \"", padding_mode_str, "\" not supported, expect zeros, border or reflection");
+  if (mode_str == "bicubic") {
+    mode_i_ = 2;
+  } else if (mode_str == "nearest") {
+    mode_i_ = 1;
+  } else {
+    mode_i_ = 0;
+  }
+  if (padding_mode_str == "reflection") {
+    padding_mode_i_ = 2;
+  } else if (padding_mode_str == "border") {
+    padding_mode_i_ = 1;
+  } else {
+    padding_mode_i_ = 0;
+  }
+}
+
+template <typename T>
+Status GridSample<T>::ComputeInternal(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  const auto& dims_input = X->Shape().GetDims();
+
+  if (dims_input.size() != 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Only 4-D tensor supported, got ", dims_input.size());
+  }
+  const Tensor* Grid = context->Input<Tensor>(1);
+  const auto& dims_grid = Grid->Shape().GetDims();
+
+  if (dims_grid.size() != 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Only 4-D tensor supported, got ", dims_input.size());
+  }
+  if (dims_grid.data()[3] != 2) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Only 4-D tensor supported, got ", dims_input.size());
+  }
+
+  std::vector<int64_t> dims_output(4);
+  dims_output[0] = dims_grid[0];
+  dims_output[1] = dims_input[1];
+  dims_output[2] = dims_grid[1];
+  dims_output[3] = dims_grid[2];
+  Tensor* Y = context->Output(0, dims_output);
+
+  typedef typename ToCudaType<T>::MappedType CudaT;
+  CudaT* Y_data = reinterpret_cast<CudaT*>(Y->MutableData<T>());
```

Comment:
nit: corner case ? return early if the output is going to be "empty" after line 76 ?

Reply:
Seems other OPs do not check the null output. @askhade does the framework do the sanity checking outside of the OP?

Reply:
Not null, just empty. If `dims_output` had a `0` in it, you could just return at this stage..

Reply:
The schema has defined the output shape. If the ort framework enforces that outside of the OP, then we do not need check again here. Otherwise it makes sense to check.

Reply:
The framework doesn't do any enforcing outside the op, the implementation is where such handling needs to lie. 

There is a rare (but not totally impossible) scenario of seeing "empty" outputs that if not handled early on will reach `GridSampleImpl()` and the `blocksPerGrid ` will be computed to be zero and the CUDA kernel will get launched with 0 blocks per grid and it will lead to the cryptic "invalid cuda configuration" error.

As an example, please see here, how we handle it in Resize/Upsample: https://github.com/microsoft/onnxruntime/blob/d0ff2621eeb2d965cfd8c5fc458d3f4615d6c16c/onnxruntime/core/providers/cuda/tensor/upsample.cc#L64